src/models/MSC3089TreeSpace.ts

```diff
@@ -0,0 +1,433 @@
+            if (startOrder && endOrder) {
+                if (startOrder === endOrder) {
+                    // Error case: just move +1 to break out of awful math
+                    newOrder = nextString(startOrder);
+                } else {
+                    newOrder = averageBetweenStrings(startOrder, endOrder);
+                }
+            } else {
+                if (startOrder) {
+                    // We're at the end (endOrder is null, so no explicit order)
+                    newOrder = nextString(startOrder);
+                } else if (endOrder) {
+                    // We're at the start (startOrder is null, so nothing before us)
+                    newOrder = prevString(endOrder);
+                } else {
+                    // Both points are unknown. We're likely in a range where all the children
+                    // don't have particular order values, so we may need to update them too.
+                    // The other possibility is there's only us as a child, but we should have
+                    // shown up in the other states.
+                    ensureBeforeIsSane = true;
+                }
+            }
+        }
+
+        if (ensureBeforeIsSane) {
+            // We were asked by the order algorithm to prepare the moving space for a landing
+            // in the undefined order part of the order array, which means we need to update the
+            // spaces that come before it with a stable order value.
+            let lastOrder: string;
+            for (let i = 0; i <= index; i++) {
+                const target = ordered[i];
+                if (i === 0) {
+                    lastOrder = target.order;
+                }
+                if (!target.order) {
+                    // XXX: We should be creating gaps to avoid conflicts
+                    lastOrder = lastOrder ? nextString(lastOrder) : DEFAULT_ALPHABET[0];
+                    const currentChild = parentRoom.currentState.getStateEvents(EventType.SpaceChild, target.roomId);
+                    const content = currentChild?.getContent() ?? { via: [this.client.getDomain()] };
+                    await this.client.sendStateEvent(parentRoom.roomId, EventType.SpaceChild, {
+                        ...content,
+                        order: lastOrder,
+                    }, target.roomId);
+                } else {
+                    lastOrder = target.order;
+                }
+            }
+            newOrder = nextString(lastOrder);
+        }
+
+        // TODO: Deal with order conflicts by reordering
+
+        // Now we can finally update our own order state
+        const currentChild = parentRoom.currentState.getStateEvents(EventType.SpaceChild, this.roomId);
+        const content = currentChild?.getContent() ?? { via: [this.client.getDomain()] };
+        await this.client.sendStateEvent(parentRoom.roomId, EventType.SpaceChild, {
+            ...content,
+            order: newOrder,
+        }, this.roomId);
```

Comment:
I managed to massively simplify this by use of a method which creates a range of strings in between a gap and using an array to diff instead of trying to loop with prev/nextString to move one-by-one.

```typescript
const reorder = <T extends any>(list: T[], startIndex: number, endIndex: number): T[] => {
    const result = Array.from(list);
    const [removed] = result.splice(startIndex, 1);
    result.splice(endIndex, 0, removed);

    return result;
};

export const midPointsBetweenStrings = (a: string, b: string, count: number, alphabet = ALPHABET): string[] => {
    const n = Math.max(a.length, b.length);
    const aBase = stringToBase(pad(a, n, alphabet), alphabet);
    const bBase = stringToBase(pad(b, n, alphabet), alphabet);
    const step = (bBase - aBase) / (count + 1);
    if (step < 1) {
        return [];
    }
    return Array(count).fill(undefined).map((_, i) => baseToString(aBase + step + (i * step), alphabet));
};

public moveRootSpace(fromIndex: number, toIndex: number): void {
    if (
        fromIndex < 0 || toIndex < 0 ||
        fromIndex > this.rootSpaces.length || toIndex > this.rootSpaces.length ||
        fromIndex === toIndex
    ) {
        return;
    }
	const orders = this.rootSpaces.map(this.getSpaceTagOrdering);
	  
	const newRootSpaces = reorder(this.rootSpaces, fromIndex, toIndex);
	let firstIncorrectIdx = -1;
	let lastIncorrectIdx = -1;
	for (let i = 0; i < this.rootSpaces.length; i++) {
	    if (this.rootSpaces[i] !== newRootSpaces[i] || (i <= toIndex && orders[i] === undefined)) {
	        if (firstIncorrectIdx < 0) firstIncorrectIdx = i;
	        lastIncorrectIdx = i;
	    }
	}
	  
	if (firstIncorrectIdx < 0) return;
	  
	const prevOrder = firstIncorrectIdx > 0
	    ? orders[firstIncorrectIdx]
	    : String.fromCharCode(ALPHABET_START).repeat(5);
	const nextOrder = orders[lastIncorrectIdx + 1]
	    ?? String.fromCharCode(ALPHABET_END).repeat(prevOrder.length + 1);
	  
	const midpoints = midPointsBetweenStrings(prevOrder, nextOrder, 1 + lastIncorrectIdx - firstIncorrectIdx);
	for (let i = firstIncorrectIdx, j = 0; i <= lastIncorrectIdx; i++, j++) {
	    this.setRootSpaceOrder(newRootSpaces[i], midpoints[j]);
	}
```